### ISSUE TITLE
PyStan 3 Must-Haves

### ISSUE THREAD

**Issue by ariddell** [stan-dev/pystan#305]
This issue tracks feature requests which are (1) highly desired and (2) depend on or are related to [planned API refactoring related to "Stan 3"](https://github.com/stan-dev/stan/wiki/User-Interface-Guidelines-for-Developers).

Must-Haves:
- [ ] New API (see [wiki page](https://github.com/stan-dev/stan/wiki/User-Interface-Guidelines-for-Developers))
- [ ] Split into a backend (httpstan) and a frontend (pystan) to make development easier for all developers (but especially those interested in adding or changing frontend features).
- [ ] Deprecate ``stan`` function (warnings need to start ASAP)
- [ ] Automatic caching of models and fits in platform-specific directory or similar (e.g., ``~/.cache/pystan/models`` on Linux) (#203, #204)
- [ ] Safe pickling of fit objects.
- [ ] Sampling in parallel via threads (i.e., stop using ``multiprocessing``) see #179 
- [ ] Get rid of need for explicit ``.extract()``, also deal with ``permuted`` issue see stan-dev/rstan#345
- [ ] Harmonize variable names across interfaces. See #216 stan-dev/stan#1801

Features/Nice-to-Haves:
- [ ] end-to-end test comparing cmdstan and pystan output. Should be identical given fixed random seeds. Make sure chain ids are the same.
- [ ] progress bars that work in Jupyter Notebook, IPython, etc.
- [ ] (developer) pytest instead of unittest
- [ ] (developer) comprehensive linting using yapf and flake8; checked on every pull request
- [ ] (developer) type checking with mypy; checked on every pull request

Additional suggestions welcome.

Edits (2018-04-27): various additions

**Comment by IamGianluca** [comment]
Hi Allen, 

In a broader sense, I would suggest:
1. Start labeling the Issues in GitHub in order to make it easier for new contributors to identify tasks they can work on;
2. I'm not sure how you feel about test runners, but most Python open-source projects have been moving over the last few years from `unittest` and `nose` to `pytest`. `pytest` is quite more flexible;
3. Treat `ADVI` and `NUTS` results in the same way. I find it strange to have to read a .csv file to work with the output of variational inference (I'm aware something like that should make it through the next release, but I haven't tested it yet). I think consistency here would be welcomed, especially if we are expecting researchers to keep adding new samplers in the future.

I absolutely agree on the automatic caching of models, but for me it would be better to save them into hidden subdirectories inside every Python package directory (similar to what Jupyter does with its cache or Python does with the byte compiled files).

**Comment by bob-carpenter** [comment]
> On Apr 8, 2017, at 7:21 PM, Gianluca Rossi <notifications@github.com> wrote:
> 
> Hi Allen,
> 
> In a broader sense, I would suggest:
> 
> 	• Start labeling the Issues in GitHub in order to make it easier for new contributors to identify tasks they can work on;

We have the label in our other projects, but haven't gone
back over everything.  It's also hard to tell what a new
contributor could do based on what their background is.  There's
complexity across the C++, math/stats, and tangled code.

> 	• I'm not sure how you feel about test runners, but most Python open-source projects have been moving over the last few years from unittest and nose to pytest. pytest is quite more flexible;
> 	• Treat ADVI and NUTS results in the same way.

That's the plan.

> I find it strange to have to read a .csv file to work with the output of variational inference (I'm aware something like that should make it through the next release, but I haven't tested it yet). I think consistency here would be welcomed, especially if we are expecting researchers to keep adding new samplers in the future.

This was an expedient to get it in.  Nobody wanted to do a deeper
integration with the refactor on the horizon.  The new architecture
means all this can go through C++, but new algorithms tend to get
cobbled together in Python/R and then it's a huge effort for the dev
team to push them through all the interfaces in a consistent way.
That's one reason we don't have many new algorithms.  If this step
were easier, we'd have max marginal likelihood, streaming VB, and
expectation propagation already.

- Bob

**Comment by ariddell** [comment]
@IamGianluca which big projects have switched to py.test? I noticed curio uses it, which is a big vote of confidence.

**Comment by IamGianluca** [comment]
Hi Allen,

`pandas` for instance just recently moved from `nose` to `pytest`, just to mention a very large project. At Mozilla they are moving most of their projects to `pytest` too. These are the slides from one of their lead developers, which nicely outline the advantages of `pytest` over `nose` or `unittest` ([slides](https://agopian.info/presentations/2015_06_djangocon_europe/)). Lots of other large projects like `pypy`, `scikit-learn`, `astropy`, `six` use it too.

In my personal experience I recently started to use it at work and I can tell you the fixture model makes your testing code way more maintainable. Coming from `nose`, I've also enormously appreciated the fact that the syntax is more pythonic (`assert <something> == <else>` vs `assert_equal(<something>, <else>)`). One advantage I often hear, but I'm not entirely sure if true it's unique to `pytest`, is that with `pytest-xdist` you can run tests in parallel. If you have a large test code base that could be very handy :-)

Speaking of `nose`, from March 2016 there have been 2 commit merged into the master branch. I'm not sure this is still in active development.

Anyway, I think all three are very good test runners :+1:

**Comment by ariddell** [comment]
Looks like @jburos and others have written some caching code: https://github.com/hammerlab/stancache . Automatic caching will be part of PyStan soon. I'm working on it today, in fact.

**Comment by riddell-stan** [comment]
Closing as this is all more or less finished. Development continues at https://github.com/stan-dev/pystan-next
